File: quick.rb
def quick_sort(array, low = 0, high = nil)
    return array if array.length < 2
    high ||= array.length - 1
    if low < high
      pivot_index = partition(array, low, high)
      quick_sort(array, low, pivot_index-1)
      quick_sort(array, pivot_index+1, high)
    end
    array
end

def partition(array, low, high)
  pivot = array[high]
  i = low - 1
  for j in low...high
    if array[j] <= pivot
      i += 1
      array[i], array[j] = array[j], array[i]
    end
  end
  array[i+1], array[high] = array[high], array[i+1]
  i+1
end

array = [38, 27, -43, 3, -9, 82, -10, 19]
p "Unsorted array: ", array
sorted_array = quick_sort(array)
p "Sorted array: ", sorted_array

File: sorting/hard/quick.js
const quickSort = (a,low=0,high=null) => {
    if (a.length < 2) return a
    high = high === null ? a.length-1 : high
    if(low < high){
        pivot_index = partition(a,low,high)
        quickSort(a,low,pivot_index-1)
        quickSort(a,pivot_index+1,high)
    }
    return a
}

const partition = (a,low,high) => {
    pivot = a[high]
    i = low -1
    for(j=low;j<high;j++){
        if(a[j] <= pivot){
            i++
            [a[i], a[j]] = [a[j], a[i]]
        }
    }
    [a[i+1], a[high]] = [a[high],a[i+1]]
    return i+1
}

a = [7,1,5,-11,60,8,-1,2]
console.log("Unsorted Array: ",a);
console.log("Sorted Array: ",quickSort(a))